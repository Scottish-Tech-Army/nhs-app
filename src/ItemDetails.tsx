import React from "react";
import { getItemTemplate } from "./data/TraumaTower";
import { useNavigate, useParams } from "react-router-dom";
import "./App.css";
import { ReactComponent as ArrowLeft } from "./icons/arrow-left.svg";

// Icons vuesax linear. Licence: https://iconsax.io/#license

function ItemDetails() {
  let { boxTemplateId, itemId } = useParams();
  let navigate = useNavigate();

  const itemTemplate =
    boxTemplateId && itemId
      ? getItemTemplate(boxTemplateId, itemId)
      : undefined;

  return (
    <div className="item-details">
      <header>
        <button
          type="button"
          className="back"
          aria-label="Back"
          onClick={() => navigate(-1)}
        >
          <ArrowLeft />
        </button>

        <h1>Item Details</h1>
      </header>
      <main>
        {itemTemplate && (
          <>
            <img
              src={`/items/${itemTemplate.photo}`}
              className="item-photo"
              alt=""
            />
            <h2 className="name">{itemTemplate.name}</h2>
            <dl className="details">
              <div className="description">
                <dt>Description</dt>
                <dd>{itemTemplate.description}</dd>
              </div>
              <div className="size">
                <dt>Size</dt>
                <dd>{itemTemplate.size || "N/A"}</dd>
              </div>
              <div className="location">
                <dt>Location</dt>
                <dd>{itemTemplate.location}</dd>
              </div>
            </dl>
          </>
        )}
      </main>
    </div>
  );
}

export default ItemDetails;
