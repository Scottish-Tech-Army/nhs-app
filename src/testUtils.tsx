import React from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import type { AppStore, RootState } from "./data/store";
import boxContentsReducer, { initialState } from "./data/BoxContentsSlice";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProvider(
  component: React.ReactElement,
  {
    preloadedState = { boxContents: initialState },
    store = configureStore({
      reducer: { boxContents: boxContentsReducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {

  return {
    store,
    ...render(<Provider store={store}>{component}</Provider>, renderOptions),
  };
}
