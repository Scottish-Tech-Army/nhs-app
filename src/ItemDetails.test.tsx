/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/prefer-screen-queries */
import React from "react";

import { getByText, screen } from "@testing-library/react";
import { Routes, Route } from "react-router-dom";

import ItemDetails from "./ItemDetails";
import { renderWithProvider } from "./testUtils";

describe("ItemDetails", () => {
  it("rendered an item details page with size", async () => {
    const { container } = renderWithRoute("trauma-chest-drain", "9");
    expect(screen.getByText("Item Details")).toBeDefined();

    expect(
      getByText(document.querySelector("h2.name")!, "Suture pack")
    ).toBeDefined();
    expect(
      getByText(document.querySelector("div.size")!, "Standard")
    ).toBeDefined();
    expect(
      getByText(
        document.querySelector("div.description")!,
        "Pack of equipment for suturing"
      )
    ).toBeDefined();
    expect(
      getByText(document.querySelector("div.location")!, "Resus store XX")
    ).toBeDefined();

    expect(document.querySelector("img.item-photo")).toHaveAttribute(
      "src",
      "/items/standardsuture.jpg"
    );

    expect(container).toMatchSnapshot();
  });

  it("rendered an item details page without size", async () => {
    const { container } = renderWithRoute("trauma-chest-drain", "11");
    expect(screen.getByText("Item Details")).toBeDefined();

    expect(
      getByText(document.querySelector("h2.name")!, "Chest drain bottle")
    ).toBeDefined();
    expect(getByText(document.querySelector("div.size")!, "N/A")).toBeDefined();
    expect(
      getByText(
        document.querySelector("div.description")!,
        "Chest drain bottle for draining air and blood"
      )
    ).toBeDefined();
    expect(
      getByText(document.querySelector("div.location")!, "Resus store XX")
    ).toBeDefined();

    expect(document.querySelector("img.item-photo")).toHaveAttribute(
      "src",
      "/items/chestdrainbottle.jpg"
    );

    expect(container).toMatchSnapshot();
  });

  it("does not render if no containerTemplateId", async () => {
    const { container } = renderWithRoute("", "3");

    expect(container).toHaveTextContent("Unknown path");
  });

  it("does not render if no itemNumber", async () => {
    const { container } = renderWithRoute("trauma-chest-drain", "");

    expect(container).toHaveTextContent("Unknown path");
  });

  it("does not render if unknown containerTemplateId", async () => {
    const { container } = renderWithRoute("Unknown", "4");

    expect(container.children).toHaveLength(1);
  });

  it("does not render if unknown itemNumber", async () => {
    const { container } = renderWithRoute("trauma-chest-drain", "Unknown");

    expect(container.children).toHaveLength(1);
  });

  it("does not render if unknown itemNumber - container number too high", async () => {
    const { container } = renderWithRoute("trauma-chest-drain", "15");

    expect(container.children).toHaveLength(1);
  });

  it("can return to previous page", async () => {
    const { user, history } = renderWithRoute("trauma-chest-drain", "3");

    await user.click(screen.getByRole("button", { name: "back" }));

    expect(history.location.pathname).toEqual(
      "/container/trauma-chest-drain/5"
    );
  });

  function renderWithRoute(containerTemplateId: string, itemId: string) {
    // Add routes to get the contents of useParams populated
    return renderWithProvider(
      <Routes>
        <Route
          path="item/:containerTemplateId/:itemId"
          element={<ItemDetails />}
        />
        <Route path="*" element={<div>Unknown path</div>} />
      </Routes>,
      {
        initialRoutes: [
          "/",
          `/container/${containerTemplateId}/5`,
          `/item/${containerTemplateId}/${itemId}`,
        ],
      }
    );
  }
});
