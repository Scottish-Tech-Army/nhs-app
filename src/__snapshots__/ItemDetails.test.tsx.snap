// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetails rendered an item details page with size 1`] = `
<div>
  <div
    class="item-details"
  >
    <header>
      <button
        aria-label="back"
        class="back"
        type="button"
      >
        <svg>
          arrow-left.svg
        </svg>
      </button>
      <h1>
        Item Details
      </h1>
    </header>
    <main>
      <img
        alt=""
        class="item-photo"
        src="/items/standardsuture.jpg"
      />
      <h2
        class="name"
      >
        Suture pack
      </h2>
      <dl
        class="details"
      >
        <div
          class="description"
        >
          <dt>
            Description
          </dt>
          <dd>
            Pack of equipment for suturing
          </dd>
        </div>
        <div
          class="size"
        >
          <dt>
            Size
          </dt>
          <dd>
            Standard
          </dd>
        </div>
        <div
          class="location"
        >
          <dt>
            Location
          </dt>
          <dd>
            Resus store XX
          </dd>
        </div>
      </dl>
    </main>
  </div>
</div>
`;

exports[`ItemDetails rendered an item details page without size 1`] = `
<div>
  <div
    class="item-details"
  >
    <header>
      <button
        aria-label="back"
        class="back"
        type="button"
      >
        <svg>
          arrow-left.svg
        </svg>
      </button>
      <h1>
        Item Details
      </h1>
    </header>
    <main>
      <img
        alt=""
        class="item-photo"
        src="/items/chestdrainbottle.jpg"
      />
      <h2
        class="name"
      >
        Chest drain bottle
      </h2>
      <dl
        class="details"
      >
        <div
          class="description"
        >
          <dt>
            Description
          </dt>
          <dd>
            Chest drain bottle for draining air and blood
          </dd>
        </div>
        <div
          class="size"
        >
          <dt>
            Size
          </dt>
          <dd>
            N/A
          </dd>
        </div>
        <div
          class="location"
        >
          <dt>
            Location
          </dt>
          <dd>
            Resus store XX
          </dd>
        </div>
      </dl>
    </main>
  </div>
</div>
`;
